cpp/dolfinx/common/IndexMap.cpp

```diff
@@ -32,6 +32,353 @@ std::array<std::vector<int>, 2> build_src_dest(MPI_Comm comm,
+  // that are ghosts on others
+  std::vector<std::int64_t> send_indices, recv_indices;
+  for (auto& d : send_data)
+    send_indices.insert(send_indices.end(), d.begin(), d.end());
+
+  recv_indices.resize(recv_disp.back());
+  ierr = MPI_Neighbor_alltoallv(send_indices.data(), send_sizes.data(),
+                                send_disp.data(), MPI_INT64_T,
+                                recv_indices.data(), recv_sizes.data(),
+                                recv_disp.data(), MPI_INT64_T, comm0);
+  dolfinx::MPI::check_error(comm, ierr);
+
+  // Free the communicator
+  ierr = MPI_Comm_free(&comm0);
+  dolfinx::MPI::check_error(comm, ierr);
+
+  // --- Step 2 ---: Create a map from the indices in `recv_indices` (i.e.
+  // indices owned by this process that are in `indices` on other processes) to
+  // their owner in the submap. This is required since not all indices in
+  // `recv_indices` will necessarily be in `indices` on this process, and thus
+  // other processes must own them in the submap.
+
+  // Create a map from (global) owned shared indices owned to processes that can
+  // own them.
+  // FIXME Avoid map
+  const int rank = dolfinx::MPI::rank(comm);
+  // A map from (global) indices in `recv_indices` to a list of processes that
+  // can own the index in the submap.
+  std::map<std::int64_t, std::vector<std::int32_t>>
+      global_idx_to_possible_owner;
+  const std::array local_range = imap.local_range();
+  for (std::size_t i = 0; i < recv_disp.size() - 1; ++i)
+  {
+    for (int j = recv_disp[i]; j < recv_disp[i + 1]; ++j)
+    {
+      std::int64_t idx = recv_indices[j];
+      // Check that the index is in the submap
+      if (idx != -1)
+      {
+        // Compute the local index
+        std::int32_t idx_local = idx - local_range[0];
+        assert(idx_local >= 0);
+        assert(idx_local < local_range[1]);
+
+        // Check if index is in the submap on this process. If so, this process
+        // remains its owner in the submap. Otherwise, add the process that sent
+        // it to the list of possible owners.
+        if (is_in_submap[idx_local])
+          global_idx_to_possible_owner[idx].push_back(rank);
+        else
+          global_idx_to_possible_owner[idx].push_back(dest[i]);
+      }
+    }
+  }
+
+  // Choose the submap owner for each index in `recv_indices`
+  std::vector<std::int32_t> send_owners;
+  send_owners.reserve(recv_indices.size());
+  for (auto idx : recv_indices)
+  {
+    // Check the index is in the submap, otherwise send -1
+    if (idx != -1)
+    {
+      // TODO Choose new owner randomly for load balancing
+      const std::vector<std::int32_t>& possible_owners
+          = global_idx_to_possible_owner[idx];
+      // const int random_index = std::rand() % possible_owners.size();
+      // send_owners.push_back(possible_owners[random_index]);
+
+      // FIXME TEMPORARILY CHOOSE THE FIRST PROCESS, MAKE RANDOM
+      send_owners.push_back(possible_owners[0]);
```

Comment:
I wouldn't make it completely random, as results would be nondeterministic. However, if you sort the owners for each index when you create `global_idx_to_possible_owner` we can get a consistent result